wasm/wat/parser.go

```diff
@@ -222,24 +283,126 @@ func (p *moduleParser) parseImportFunc(tok tokenType, tokenBytes []byte, _, _ ui
 			return fmt.Errorf("redundant name: %s", name)
 		}
 		fn.funcName = name
-	case tokenLParen:
-		return errors.New("TODO: handle (type..) and inlined type (param..) (result..)")
-		// * typeuse https://www.w3.org/TR/wasm-core-1/#text-typeuse
-		// * inlined type https://www.w3.org/TR/wasm-core-1/#abbreviations%E2%91%A6
+	case tokenLParen: // start fields, ex. (param or (result
+		p.tokenParser = p.beginField
+		return nil
 	case tokenRParen: // end of this import func
-		// TODO: once we handle import types, this won't always be empty
-		if len(p.module.types) == 0 {
-			p.module.types = append(p.module.types, typeFuncEmpty)
+		if p.currentType == nil {
+			p.currentType = typeFuncEmpty
 		}
-		p.module.importFuncs[len(p.module.importFuncs)-1].typeIndex = 0
-		p.currentField = fieldModuleImport
-		p.tokenParser = p.parseImport
+
+		fn := p.module.importFuncs[len(p.module.importFuncs)-1]
+
+		// search for an existing signature that matches the current type
+		for i, t := range p.module.types {
+			if bytes.Equal(p.currentType.params, t.params) && bytes.Equal(p.currentType.results, t.results) {
+				fn.typeIndex = uint32(i)
+				p.currentType = nil
+				break
+			}
+		}
+
+		// if we didn't find a match, we need to insert a new type and use it
+		if p.currentType != nil { // new type
+			fn.typeIndex = uint32(len(p.module.types))
+			p.module.types = append(p.module.types, p.currentType)
+			p.currentType = nil
+		}
+
+		// reset parsing state
+		p.currentValue0, p.currentValue1 = nil, nil
+		p.currentParamIndex = 0
+		return p.endField()
+	default:
+		return p.unexpectedToken(tok, tokenBytes)
+	}
+	return nil
+}
+
+func (p *moduleParser) parseImportFuncParam(tok tokenType, tokenBytes []byte, _, _ uint32) error {
+	switch tok {
+	case tokenID: // Ex. $len
+		return errors.New("TODO param name ex (param $len i32), but not in abbreviation ex (param $x i32 $y i32)")
+	case tokenKeyword: // Ex. i32
+		vt, err := parseValueType(tokenBytes)
+		if err != nil {
+			return err
+		}
+
+		// Here, we have to check for the abbreviated form, ex. (param i32 i32), noting this does not allow names!
+		if p.currentParamIndex > 0 && p.currentParams != nil { // ex. (param i32) (param i32 i32)
+			return errors.New("multiple type abbreviation is unexpected with multiple params")
+		}
+		p.currentParams = append(p.currentParams, vt)
+	case tokenRParen: // end of this field
+		if p.currentParams == nil {
+			return errors.New("expected a type")
+		}
+
+		// add the parameters to the current type
+		if p.currentType == nil {
+			p.currentType = &typeFunc{}
+		}
+		p.currentType.params = append(p.currentType.params, p.currentParams...)
+
+		// since multiple param fields are valid, prepare for any next.
```

Comment:
sorry what do you mean by "multiple param fields"? any example?

Reply:
good question!

```suggestion
		// since multiple param fields are valid, ex `(func (param i32) (param i64)`, prepare for any next.
```